app/models/allocation.rb

```diff
@@ -50,22 +50,41 @@ class Allocation < VersionedModel
 
   attribute :complex_cases, Types::JSONB.new(Allocation::ComplexCaseAnswers)
 
-  def refresh_moves_count!
-    self.moves_count = moves.not_cancelled.count
-    save!
-  end
+  after_initialize :initialize_state
+  delegate :fill, :unfill, :cancel, to: :state_machine
 
-  def cancel
+  def cancel_with_moves
     comment = 'Allocation was cancelled'
 
     assign_attributes(
-      status: ALLOCATION_STATUS_CANCELLED,
       cancellation_reason: CANCELLATION_REASON_OTHER,
       cancellation_reason_comment: comment,
       moves: moves.each { |move| move.cancel(comment: comment) },
       moves_count: 0,
     )
+    cancel
 
     save!
   end
+
+  def refresh_status_and_moves_count!
+    moves.not_cancelled.empty? || moves.not_cancelled.unfilled? ? unfill : fill
```

Comment:
this looks like it will mark all allocations with at least one cancelled move as "filled". Is that the correct behaviour ?

Reply:
oh my bad, I misread the statement. Might be good to add an explanatory comment though to avoid any confusion